File: isValidBST.cpp
class Solution {
public:

    bool check(TreeNode* node, long long lb, long long ub){
           if(node == NULL)return true;

           if(node->val >= ub || node->val <= lb)return false;

           return check(node->left, lb, node->val) and check(node->right, node->val, ub);

    }
    bool isValidBST(TreeNode* root) {
          return check(root, LLONG_MIN, LLONG_MAX);
    }
};

File: isValidBST.py
class Solution:
    def isValidBST(self, root):
        return self.check(root, float('-inf'), float('inf'))

    def check(self, node, lb, ub):
        if node is None:
            return True

        if not lb < node.val < ub:
            return False

        return self.check(node.left, lb, node.val) and self.check(node.right, node.val, ub)
